### DISCUSSION TITLE
Hooks to post process rendered HTML

### DISCUSSION THREAD

**Discussion by jonluca** [vercel/next.js#62777] [Help]
### Summary

I'd like to have more control over the post processing of html - for instance, conditionally running my own CSS optimizer (instead of critters) based on the RequestContext.

Right now `postProcessHTML` from `packages/next/src/server/post-process.ts` is called before every `RenderResult` and only applies the hard coded options. 

Is there a place where we can hook into the rendered HTML and modify it before it's sent along?

**Comment by jonluca** [comment]
I recognize I can do this with a proxy or nginx, but I'm wondering if theres a supported way of doing this natively

**Comment by lukajincharadze** [comment]
Create a custom server file in your Next.js project directory. You can name it server.js:
`// server.js

const { createServer } = require('http');
const { parse } = require('url');
const next = require('next');

const dev = process.env.NODE_ENV !== 'production';
const app = next({ dev });
const handle = app.getRequestHandler();

app.prepare().then(() => {
  createServer((req, res) => {
    const parsedUrl = parse(req.url, true);
    const { pathname, query } = parsedUrl;

    if (pathname === '/custom-route') {
      // Modify the rendered HTML here
      const modifiedHtml = '<html><body><h1>Hello, World!</h1></body></html>';
      
      // Send the modified HTML response
      res.setHeader('Content-Type', 'text/html');
      res.end(modifiedHtml);
    } else {
      handle(req, res, parsedUrl);
    }
  }).listen(3000, (err) => {
    if (err) throw err;
    console.log('> Ready on http://localhost:3000');
  });
});
`

Run the custom server instead of the default Next.js server:
`node server.js`

and then navigate to http://localhost:3000/custom-route in your browser to see the modified HTML.

**Reply by jonluca** [reply]
This doesn't work for serverless deployments or edge runtime though

**Reply by lukajincharadze** [reply]
try this then:


```
import React from 'react';

export default function Example({ html }) {
  // Render the modified HTML
  return <div dangerouslySetInnerHTML={{ __html: html }} />;
}

export async function getServerSideProps(context) {
  // Fetch data or perform any necessary operations
  const data = await fetchData();

  // Modify the HTML based on the request context
  const modifiedHtml = modifyHtml(data);

  // Pass the modified HTML as a prop to the component
  return {
    props: {
      html: modifiedHtml,
    },
  };
}

```

**Reply by jonluca** [reply]
`getServerSideProps` does not run after final html serialization, it runs well before it. Theres the entire SSR prepass, then theres the Head injection, and then multiple other steps.

**Reply by lukajincharadze** [reply]
ohh i got it, this should work:

```
// Example component
import React, { useEffect, useState } from 'react';

const Example = ({ initialHtml }) => {
  const [html, setHtml] = useState(initialHtml);

  useEffect(() => {
    // Modify the HTML here as needed
    const modifiedHtml = modifyHtml(html);
    setHtml(modifiedHtml);
  }, [html]);

  return <div dangerouslySetInnerHTML={{ __html: html }} />;
};

export default Example;
```

**Reply by jonluca** [reply]
No that does not either - are you generating these via ChatGPT? These are obviously incorrect and fundamentally different from the way this framework operates.

The functionality I'm referencing in the original discussions does not exist, as far as I can tell - you can go in and look at the source code and see theres no hooks (afaict) for the `RenderResult` library-internal concept.

I appreciate attempting to be helpful but please make sure you understand the point of discussion first.
